### PR TITLE
Add new array deficit recorders.

### PR DESCRIPTION
**Pull request by jetuk**
- NumpyArrayNodeDeficitRecorder - timeseries of deficit.
- NumpyArrayNodeSuppliedRatioRecorder - timeseries of supply / demand
- NumpyArrayNodeCurtailmentRatioRecorder - timeseries of 1 - supply / demand

### CODE REVIEW

File: pywr/recorders/_recorders.pyx [hunk 1]
```diff
@@ -479,6 +479,88 @@ cdef class NumpyArrayNodeRecorder(NodeRecorder):
...
+        for scenario_index in self.model.scenarios.combinations:
+            max_flow = node.get_max_flow(scenario_index)
+            self._data[ts._index,scenario_index.global_id] = node._flow[scenario_index.global_id] / max_flow
+        return 0
+
+
+cdef class NumpyArrayNodeCurtailmentRatioRecorder(NumpyArrayNodeRecorder):
+    """Recorder for timeseries of curtailment ratio from a `Node`.
+
+    This class stores deficit from a specific node for each time-step of a simulation. The
```
Comment (Batch21) [inline pywr/recorders/_recorders.pyx]: First sentence here and for ratio recorder could be updated to be more relevant to the specific recorder. Maybe also worth putting the equation used to calculate the metric in the docstring

File: pywr/recorders/_recorders.pyx [hunk 2]
```diff
@@ -479,6 +479,88 @@ cdef class NumpyArrayNodeRecorder(NodeRecorder):
...
+    """
+    cpdef after(self):
+        cdef double max_flow
+        cdef ScenarioIndex scenario_index
+        cdef Timestep ts = self.model.timestepper.current
+        cdef Node node = self._node
+        for scenario_index in self.model.scenarios.combinations:
+            max_flow = node.get_max_flow(scenario_index)
+            self._data[ts._index,scenario_index.global_id] = 1 - node._flow[scenario_index.global_id] / max_flow
+
```
Comment (Batch21) [inline pywr/recorders/_recorders.pyx]: Need to call .register on all 3 recorders

File: pywr/recorders/_recorders.pxd [hunk 3]
```diff
@@ -1,5 +1,5 @@
 from pywr._component cimport Component
-from pywr._core cimport Timestep, AbstractNode, AbstractStorage, Storage, ScenarioIndex, Scenario
+from pywr._core cimport Timestep, AbstractNode, Node, AbstractStorage, Storage, ScenarioIndex, Scenario
```
Comment (Batch21) [inline pywr/recorders/_recorders.pxd]: Is this needed?

Comment (jetuk) [reply pywr/recorders/_recorders.pxd]: `cdef Node node = self._node`

It's needed for this type definition I think.

File: pywr/recorders/_recorders.pyx (hunk 2 above)
Comment (jetuk) [reply pywr/recorders/_recorders.pyx]: Done

File: pywr/recorders/_recorders.pyx (hunk 1 above)
Comment (jetuk) [reply pywr/recorders/_recorders.pyx]: I've made a few changes to the docstrings including this one. I added a "notes" and "see also" section to each. The notes contain the equation that each recorder performs.

**Comment by jetuk** [comment]
@Batch21 thanks for comments. I've addressed them in the last commit (see above).

**Pull request merged**